src/StoreApi/Routes/V1/ProductsBySlug.php

```diff
@@ -0,0 +1,115 @@
+<?php
+namespace Automattic\WooCommerce\StoreApi\Routes\V1;
+
+use Automattic\WooCommerce\StoreApi\Exceptions\RouteException;
+
+/**
+ * ProductsBySlug class.
+ */
+class ProductsBySlug extends AbstractRoute {
+	/**
+	 * The route identifier.
+	 *
+	 * @var string
+	 */
+	const IDENTIFIER = 'products-by-slug';
+
+	/**
+	 * The routes schema.
+	 *
+	 * @var string
+	 */
+	const SCHEMA_TYPE = 'product';
+
+	/**
+	 * Get the path of this REST route.
+	 *
+	 * @return string
+	 */
+	public function get_path() {
+		return '/products/(?P<slug>[\S]+)';
+	}
+
+	/**
+	 * Get method arguments for this REST route.
+	 *
+	 * @return array An array of endpoints.
+	 */
+	public function get_args() {
+		return [
+			'args'   => array(
+				'slug' => array(
+					'description' => __( 'Slug of the resource.', 'woo-gutenberg-products-block' ),
+					'type'        => 'string',
+				),
+			),
+			[
+				'methods'             => \WP_REST_Server::READABLE,
+				'callback'            => [ $this, 'get_response' ],
+				'permission_callback' => '__return_true',
+				'args'                => array(
+					'context' => $this->get_context_param(
+						array(
+							'default' => 'view',
+						)
+					),
+				),
+			],
+			'schema' => [ $this->schema, 'get_public_item_schema' ],
+		];
+	}
+
+	/**
+	 * Get a single item.
+	 *
+	 * @throws RouteException On error.
+	 * @param \WP_REST_Request $request Request object.
+	 * @return \WP_REST_Response
+	 */
+	protected function get_route_response( \WP_REST_Request $request ) {
+		$object = $this->get_product_by_slug( $request['slug'] );
```

Comment:
Perhaps we may want to sanitize the input here? `sanitize_title()` maybe (not tested)?